AIMA4e Search

samagra14 opened the pull request:
This PR adds the generic search algorithm to the AIMA4e branch. For testing, the generic search is extended as bfs and dfs. It is then applied to the Romania problem.

samagra14 committed:
Adds generic Search

samagra14 committed:
Adds Breadth first Search as a test case

samagra14 committed:
Adds depth first search version of generic search

samagra14 commented:
@norvig Please have a look.
 Should I remove Tree search and graph search from AIMA4e or should they coexist alongside generic search?

samagra14 committed:
Adds bfs and successor functions

samagra14 committed:
Adds uniformCost Search

samagra14 committed:
Adds depth limited search and iterative deepening search

samagra14 committed:
Adds comments

samagra14 committed:
Adds comments

samagra14 committed:
Adds methods to retrieve actions

samagra14 committed:
Adds generator for actions

samagra14 committed:
Adds bfs tests

samagra14 committed:
Add tests for IDS and UCS

norvig commented on core/src/main/java/aima/core/search/basic/uninformedsearch/BreadthFirstSearch.java:
```diff
@@ -10,31 +10,79 @@
         ((LinkedList<Node>) frontier).add(nodeFactory.newRootNode(problem.initialState()));
+        // reached ← a set of states; initially empty
         HashSet<S> reached = new HashSet<>();
+        // solution ← failure
```
I don't like this style, where the comments duplicate the code. I think it is good to have the pseudocode in a block comment, but then the Java implementation should stand by itself. Any Java programmer knows what `reached = new HashSet<>();` means; we don't need to duplicate that in the comments.

samagra14 replied:
@norvig 
Sure, sir, I will have these duplicate comments removed. I added them just because of the older versions. I will have them removed.

samagra14 committed:
Adds A* search and greedy best first search

samagra14 committed:
Adds simulated Annealing and Hill climbing search

samagra14 committed:
Improvements as per test cases

samagra14 committed:
Removes unwanted comments

samagra14 commented:
@norvig 
This PR concludes all the modified algorithms.
1. I have removed a lot of unwanted interfaces(for eg SearchController etc).
2. Implemented informed search from scratch as opposed to extending uniform cost search.
3. AndOrGraphSearch and the algorithms after that have not been modified in the draft and hence I have not modified their implementation. All other algorithms have been modified.
5 Removed unwanted comments.
6. In a few algorithms, I have kept the older testSuite as it was quite exhaustive.
Please review this PR.

samagra14 commented:
@norvig 
Is there anything else that needs to be corrected and/or added in this PR?

norvig commented:
Looks good!
